front/lib/api/assistant/conversation.ts

```diff
@@ -200,126 +205,210 @@ async function renderUserMessage(
+            configurationId: m.agentConfigurationId,
+          };
         }
-      : null,
-    mentions: mentions.map((m) => {
-      if (m.agentConfigurationId) {
-        return {
-          configurationId: m.agentConfigurationId,
-        };
-      }
-      if (m.user) {
-        return {
-          provider: m.user.provider,
-          providerId: m.user.providerId,
-        };
-      }
-      throw new Error("Unreachable: mention must be either agent or user");
-    }),
-    content: userMessage.content,
-    context: {
-      username: userMessage.userContextUsername,
-      timezone: userMessage.userContextTimezone,
-      fullName: userMessage.userContextFullName,
-      email: userMessage.userContextEmail,
-      profilePictureUrl: userMessage.userContextProfilePictureUrl,
-    },
-  };
+        if (m.user) {
+          return {
+            provider: m.user.provider,
+            providerId: m.user.providerId,
+          };
+        }
+        throw new Error("Unreachable: mention must be either agent or user");
+      }),
+      content: message.userMessage.content,
+      context: {
+        username: message.userMessage.userContextUsername,
+        timezone: message.userMessage.userContextTimezone,
+        fullName: message.userMessage.userContextFullName,
+        email: message.userMessage.userContextEmail,
+        profilePictureUrl: message.userMessage.userContextProfilePictureUrl,
+      },
+    };
+    return { m, rank: message.rank, version: message.version };
+  });
 }
 
-async function renderAgentMessage(
+async function batchRenderAgentMessages(
   auth: Authenticator,
-  {
-    message,
-    agentMessage,
-    messages,
-  }: { message: Message; agentMessage: AgentMessage; messages: Message[] }
-): Promise<AgentMessageType> {
-  const [agentConfiguration, agentRetrievalAction, agentDustAppRunAction] =
+  messages: Message[]
+) {
+  if (messages.find((m) => !m.agentMessage)) {
+    throw new Error(
+      "Unreachable: batchRenderAgentMessages must be called with only agent messages"
+    );
+  }
+
+  const [agentConfigurations, agentRetrievalActions, agentDustAppRunActions] =
     await Promise.all([
-      getAgentConfiguration(auth, agentMessage.agentConfigurationId),
       (async () => {
-        if (agentMessage.agentRetrievalActionId) {
-          return await renderRetrievalActionByModelId(
-            agentMessage.agentRetrievalActionId
-          );
-        }
-        return null;
+        const agentConfigurationIds: string[] = messages.reduce(
+          (acc: string[], m) => {
+            const agentId = m.agentMessage?.agentConfigurationId;
+            if (agentId && !acc.includes(agentId)) {
+              acc.push(agentId);
+            }
+            return acc;
+          },
+          []
+        );
+        const agents = (
+          await Promise.all(
+            agentConfigurationIds.map(async (agentConfigId) => {
+              return await getAgentConfiguration(auth, agentConfigId);
```

Comment:
I took a shortcut here, WDYT? I think it's worth shipping with it and iterate if we want to optimize again.